Add groups to RegExp.ExecResult

vhiairrassary opened the pull request:
Add `groups` to `RegExp.ExecResult` (added in [ECMAScript 2018](https://ecma-international.org/ecma-262/9.0/), original [proposal](https://github.com/tc39/proposal-regexp-named-groups))

Tested in
- browsers: here Chrome 87 but similar results in Firefox 84 and Safari 13.1
<img width="671" alt="Screenshot 2020-12-29 at 09 18 03" src="https://user-images.githubusercontent.com/6972399/103269840-eaf91500-49b6-11eb-85f8-df967a196597.png">

- Node.js 12
<img width="603" alt="Screenshot 2020-12-29 at 09 17 50" src="https://user-images.githubusercontent.com/6972399/103269864-019f6c00-49b7-11eb-97d8-f4e1e3e38097.png">

⚠️ Questions:
- [Contributing guidelines](https://github.com/scala-js/scala-js/blob/master/CONTRIBUTING.md) mentions `All code contributed to the user-facing standard library (the library/ directory) should come accompanied with documentation`. Should I do it as other members of `ExecResult` are not documented?
- I was not able to find a place to add a unit test, can you point me to the correct location please?

vhiairrassary commented:
CLA signed

vhiairrassary commented:
Tests are failing due to binary incompatibility:
```
[info] Scala.js library: found 2 potential binary incompatibilities while checking against org.scala-js:scalajs-library_2.11:1.3.1 
[error]  * abstract method groups()scala.scalajs.js.| in interface scala.scalajs.js.RegExp#ExecResult is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("scala.scalajs.js.RegExp#ExecResult.groups")
[error]  * abstract method groups_=(scala.scalajs.js.|)Unit in interface scala.scalajs.js.RegExp#ExecResult is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("scala.scalajs.js.RegExp#ExecResult.groups_=")
[error] java.lang.RuntimeException: Scala.js library: Binary compatibility check failed!
[error] 	at scala.sys.package$.error(package.scala:26)
[error] 	at com.typesafe.tools.mima.plugin.SbtMima$.reportModuleErrors(SbtMima.scala:114)
[error] 	at com.typesafe.tools.mima.plugin.MimaPlugin$.$anonfun$mimaReportSettings$11(MimaPlugin.scala:78)
[error] 	at com.typesafe.tools.mima.plugin.MimaPlugin$.$anonfun$mimaReportSettings$11$adapted(MimaPlugin.scala:61)
[error] 	at scala.collection.immutable.Map$Map1.foreach(Map.scala:125)
[...]
[error] (library2_11 / mimaReportBinaryIssues) Scala.js library: Binary compatibility check failed!
```
Is there anything to do on my side?

sjrd commented:
FYI, I'm currently on vacation. I'll have a look within 10 days.

vhiairrassary commented:
Thank you 👍 Enjoy your holidays

sjrd commented on library/src/main/scala/scala/scalajs/js/RegExp.scala:
```diff
@@ -117,5 +117,6 @@ object RegExp extends js.Object {
   trait ExecResult extends js.Array[js.UndefOr[String]] {
     var index: Int
     var input: String
+    var groups: js.UndefOr[js.Dictionary[String]]
```
A named group that does not capture anything will be present but its value will be undefined:
```js
C:\Users\sjrdo>node
Welcome to Node.js v14.15.0.
Type ".help" for more information.
> var r = new RegExp("foo(?<prop>bar)?baz")
undefined
> r.exec("foobarbaz")
[
  'foobarbaz',
  'bar',
  index: 0,
  input: 'foobarbaz',
  groups: [Object: null prototype] { prop: 'bar' }
]
> r.exec("foobaz")
[
  'foobaz',
  undefined,
  index: 0,
  input: 'foobaz',
  groups: [Object: null prototype] { prop: undefined }
]
```
therefore, the typing should be:
```suggestion
    var groups: js.UndefOr[js.Dictionary[js.UndefOr[String]]]
```

sjrd replied:
Moreover, since this property was added in ES 2018, it should receive a proper Scaladoc comment with an ES 2018 badge. See for example
https://github.com/scala-js/scala-js/blob/814d4d99cecaf351cf22ed365efb5d35c1f59ee2/library/src/main/scala/scala/scalajs/js/BigInt.scala#L18
You'll also need to add the CSS class in
https://github.com/scala-js/scala-js/blob/814d4d99cecaf351cf22ed365efb5d35c1f59ee2/assets/additional-doc-styles.css#L1

vhiairrassary replied:
Fixed

sjrd reviewed (changes requested):
Here I am at last. Sorry for the delay.

> Contributing guidelines mentions All code contributed to the user-facing standard library (the library/ directory) should come accompanied with documentation. Should I do it as other members of ExecResult are not documented?

It's not critical in this case, but if you can add some documentation, that's always better.

> I was not able to find a place to add a unit test, can you point me to the correct location please?

The correct place would be to add a new file `RegExpTest.scala` in https://github.com/scala-js/scala-js/tree/master/test-suite/js/src/test/scala/org/scalajs/testsuite/library . You can take inspiration from the file `ObjectTest.scala` in the same directory.

> Tests are failing due to binary incompatibility:
> [...]
> Is there anything to do on my side?

Yes. You'll need to add an exclusion in https://github.com/scala-js/scala-js/blob/814d4d99cecaf351cf22ed365efb5d35c1f59ee2/project/BinaryIncompatibilities.scala#L31-L34, with the justification that is a "New optional member in JS trait, not an issue.".

vhiairrassary commented:
Thank you for this detailed review, I addressed all the points except the documentation as I didn't found something good on [MDN website](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/exec).

sjrd reviewed (approved):
Excellent. Very nice. Thank you, also for paying attention to all the details, notably the coding style.
